File: example-microservice-java/src/main/java/com/example/java/customer/repository/CustomerRepositoryImpl.java
package com.example.java.customer.repository;

import com.example.java.customer.api.dto.CustomerCreateDto;
import com.example.java.customer.api.dto.CustomerUpdateDto;
import com.example.java.customer.service.CustomerModel;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Repository;

import java.util.Optional;
import java.util.UUID;

@Slf4j
@Repository
@AllArgsConstructor
public class CustomerRepositoryImpl implements CustomerRepository {

    private CustomerJpaRepository customerJpaRepository;

    private static final CustomerMapper customerMapper = CustomerMapper.INSTANCE;

    public Optional<CustomerModel> getCustomer(UUID id) {
        return customerJpaRepository.findById(id)
                .map(customerMapper::customerEntityToModel);
    }

    @Override
    public CustomerModel createCustomer(CustomerCreateDto newCustomer) {
        CustomerEntity createdCustomerEntity = customerMapper.customerCreateDtoToEntity(newCustomer);
        return customerMapper.customerEntityToModel(customerJpaRepository.save(createdCustomerEntity));
    }

    @Override
    public CustomerModel updateCustomer(UUID id, CustomerUpdateDto updatedCustomer) {
        CustomerEntity updatedCustomerEntity = customerMapper.customerUpdateDtoToEntity(updatedCustomer);
        return customerMapper.customerEntityToModel(customerJpaRepository.save(updatedCustomerEntity));
    }

    @Override
    public void deleteCustomer(UUID id) {
        customerJpaRepository.deleteById(id);
    }

}

File: example-microservice-kotlin/src/main/kotlin/com/gooddata/kotlin/customer/repository/CustomerRepositoryImpl.kt
package com.example.kotlin.customer.repository

import com.example.kotlin.customer.api.dto.CustomerCreateDto
import com.example.kotlin.customer.api.dto.CustomerUpdateDto
import com.example.kotlin.customer.service.CustomerModel
import org.springframework.stereotype.Repository
import java.util.*

@Repository
class CustomerRepositoryImpl(
    private val customerJpaRepository: CustomerJpaRepository
) : CustomerRepository {

    private val customerMapper = CustomerMapper.INSTANCE

    override fun getCustomer(id: UUID): CustomerModel? {
        return customerJpaRepository.findById(id)
            .map(customerMapper::customerEntityToModel)
            .orElse(null)
    }

    override fun createCustomer(newCustomer: CustomerCreateDto): CustomerModel {
        val createdCustomerEntity = customerMapper.customerCreateDtoToEntity(newCustomer)
        return customerMapper.customerEntityToModel(customerJpaRepository.save(createdCustomerEntity))
    }

    override fun updateCustomer(id: UUID, updatedCustomer: CustomerUpdateDto): CustomerModel {
        val updatedCustomerEntity = customerMapper.customerUpdateDtoToEntity(updatedCustomer)
        return customerMapper.customerEntityToModel(customerJpaRepository.save(updatedCustomerEntity))
    }

    override fun deleteCustomer(id: UUID) {
        customerJpaRepository.deleteById(id)
    }

}
